File: SwiftUI/EdgeStudio/Data/QuickstartDownloadService.swift
import Foundation
import Observation

// MARK: - QuickstartProject Model

struct QuickstartProject: Identifiable {
    let id = UUID()
    let name: String
    let directoryName: String
    let path: URL
    var isConfigured: Bool
}

// MARK: - QuickstartDownloadService

@Observable
final class QuickstartDownloadService {
    // MARK: - State

    var isDownloading = false
    var downloadProgress = 0.0
    var projects: [QuickstartProject] = []

    // MARK: - Static Constants

    static let zipURL = URL(string: "https://github.com/getditto/quickstart/archive/refs/heads/main.zip")!
    static let extractedFolderName = "quickstart-main"

    static let projectDisplayNames: [String: String] = [
        "android-java": "Android Java",
        "android-kotlin": "Android Kotlin",
        "cpp-tui": "C++ TUI",
        "dotnet-maui": ".NET MAUI",
        "dotnet-tui": ".NET TUI",
        "dotnet-winforms": ".NET WinForms",
        "edge-server": "Edge Server",
        "flutter_app": "Flutter",
        "go-tui": "Go TUI",
        "java-server": "Java Server",
        "javascript-tui": "JavaScript TUI",
        "javascript-web": "JavaScript Web",
        "kotlin-multiplatform": "Kotlin Multiplatform",
        "react-native": "React Native",
        "react-native-expo": "React Native Expo",
        "rust-tui": "Rust TUI",
        "swift": "Swift"
    ]

    // MARK: - Errors

    enum QuickstartError: LocalizedError {
        case extractionFailed(String)

        var errorDescription: String? {
            switch self {
            case let .extractionFailed(detail):
                return "Extraction failed: \(detail)"
            }
        }
    }

    // MARK: - Download & Extract

    /// Downloads the quickstart zip from GitHub, extracts it to the given destination, and returns the extracted folder URL.
    func downloadAndExtract(to destination: URL) async throws -> URL {
        await MainActor.run {
            isDownloading = true
            downloadProgress = 0.0
        }

        defer {
            Task { @MainActor in
                isDownloading = false
            }
        }

        Log.info("QuickstartDownloadService: Starting download from \(Self.zipURL)")

        await MainActor.run {
            downloadProgress = 0.1
        }

        // Download zip
        let (localZipURL, _) = try await URLSession.shared.download(from: Self.zipURL)

        Log.info("QuickstartDownloadService: Download complete, extracting to \(destination.path)")

        await MainActor.run {
            downloadProgress = 0.5
        }

        // Extract zip using /usr/bin/unzip
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip")
        process.arguments = ["-o", localZipURL.path, "-d", destination.path]

        let outputPipe = Pipe()
        let errorPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = errorPipe

        try process.run()
        process.waitUntilExit()

        let exitCode = process.terminationStatus
        if exitCode != 0 {
            let errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()
            let errorMessage = String(data: errorData, encoding: .utf8) ?? "Unknown error"
            Log.error("QuickstartDownloadService: Extraction failed with exit code \(exitCode): \(errorMessage)")
            throw QuickstartError.extractionFailed("Exit code \(exitCode): \(errorMessage)")
        }

        // Clean up temp zip
        try? FileManager.default.removeItem(at: localZipURL)

        await MainActor.run {
            downloadProgress = 1.0
        }

        let extractedPath = destination.appendingPathComponent(Self.extractedFolderName)
        Log.info("QuickstartDownloadService: Extracted to \(extractedPath.path)")
        return extractedPath
    }

    // MARK: - Configure .env Files

    /// Generates .env content and writes it to three locations: repo root, flutter_app/, and go-tui/.
    func configureEnvFiles(
        in repoRoot: URL,
        databaseId: String,
        token: String,
        authUrl: String,
        websocketUrl: String
    ) {
        // Build env content without leading whitespace on each line
        let envLines = [
            "#!/usr/bin/env bash",
            "",
            "# Auto-configured by Edge Studio",
            "DITTO_APP_ID=\"\(databaseId)\"",
            "DITTO_PLAYGROUND_TOKEN=\"\(token)\"",
            "DITTO_AUTH_URL=\"\(authUrl)\"",
            "DITTO_WEBSOCKET_URL=\"\(websocketUrl)\""
        ]
        let envContent = envLines.joined(separator: "\n")

        let envLocations: [URL] = [
            repoRoot.appendingPathComponent(".env"),
            repoRoot.appendingPathComponent("flutter_app").appendingPathComponent(".env"),
            repoRoot.appendingPathComponent("go-tui").appendingPathComponent(".env")
        ]

        let fileManager = FileManager.default

        for envURL in envLocations {
            let directory = envURL.deletingLastPathComponent()
            if fileManager.fileExists(atPath: directory.path) {
                do {
                    try envContent.write(to: envURL, atomically: true, encoding: .utf8)
                    Log.info("QuickstartDownloadService: Wrote .env to \(envURL.path)")
                } catch {
                    Log.warning("QuickstartDownloadService: Failed to write .env to \(envURL.path): \(error.localizedDescription)")
                }
            } else {
                Log.info("QuickstartDownloadService: Skipping .env for non-existent directory \(directory.path)")
            }
        }
    }

    // MARK: - Configure Edge Server YAML

    /// Generates quickstart_config.yaml in the edge-server/ subdirectory.
    func configureEdgeServerYaml(
        in repoRoot: URL,
        databaseId: String,
        token: String,
        authUrl: String
    ) throws {
        // Build YAML content without leading whitespace on each line
        let yamlLines = [
            "resources:",
            "  my_ditto_db:",
            "    resource_type: DittoDatabase",
            "    db_id: \"\(databaseId)\"",
            "    device_name: \"edge-studio-quickstart\"",
            "    subscriptions:",
            "      - \"SELECT * FROM tasks\"",
            "    auth:",
            "      server:",
            "        access_token: \"\(token)\"",
            "        auth_url: \"\(authUrl)\"",
            "        provider: \"__playgroundProvider\"",
            "  my_http_server:",
            "    resource_type: HttpServer",
            "    listen_addr: \"0.0.0.0:8080\"",
            "    databases:",
            "      my_db:",
            "        db_id: \"\(databaseId)\"",
            "        base_path: my_server",
            "        http_api: true"
        ]
        let yamlContent = yamlLines.joined(separator: "\n")

        let edgeServerDir = repoRoot.appendingPathComponent("edge-server")
        let yamlURL = edgeServerDir.appendingPathComponent("quickstart_config.yaml")

        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: edgeServerDir.path) else {
            Log.info("QuickstartDownloadService: edge-server directory not found, skipping YAML config")
            return
        }

        try yamlContent.write(to: yamlURL, atomically: true, encoding: .utf8)
        Log.info("QuickstartDownloadService: Wrote edge-server YAML to \(yamlURL.path)")
    }

    // MARK: - Discover Projects

    /// Scans the directory for known project folders and populates the projects array.
    func discoverProjects(in directory: URL, isConfigured: Bool) {
        let fileManager = FileManager.default
        var discovered: [QuickstartProject] = []

        for (dirName, displayName) in Self.projectDisplayNames {
            let projectURL = directory.appendingPathComponent(dirName)
            if fileManager.fileExists(atPath: projectURL.path) {
                let project = QuickstartProject(
                    name: displayName,
                    directoryName: dirName,
                    path: projectURL,
                    isConfigured: isConfigured
                )
                discovered.append(project)
            }
        }

        // Sort by display name for consistent ordering
        discovered.sort { $0.name < $1.name }

        Task { @MainActor in
            projects = discovered
        }

        Log.info("QuickstartDownloadService: Discovered \(discovered.count) projects in \(directory.path)")
    }

    // MARK: - Existing Folder Check

    /// Returns the URL of the quickstart-main folder if it exists at the given location.
    func existingQuickstartFolder(in location: URL) -> URL? {
        let folderURL = location.appendingPathComponent(Self.extractedFolderName)
        if FileManager.default.fileExists(atPath: folderURL.path) {
            return folderURL
        }
        return nil
    }

    // MARK: - Remove Existing Folder

    /// Removes the directory at the given URL.
    func removeExistingFolder(at url: URL) throws {
        try FileManager.default.removeItem(at: url)
        Log.info("QuickstartDownloadService: Removed folder at \(url.path)")
    }
}

File: dotnet/src/EdgeStudio.Shared/Services/QuickstartDownloadService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;

namespace EdgeStudio.Shared.Services;

/// <summary>
/// Represents a discovered quickstart project directory.
/// </summary>
public record QuickstartProject(string Name, string DirectoryName, string Path, bool IsConfigured);

/// <summary>
/// Handles downloading the Ditto quickstart repo zip, extracting it,
/// generating .env files, generating edge-server YAML config, and discovering quickstart projects.
/// </summary>
public class QuickstartDownloadService
{
    // MARK: - Constants

    public const string ZipUrl = "https://github.com/getditto/quickstart/archive/refs/heads/main.zip";
    public const string ExtractedFolderName = "quickstart-main";

    public static readonly Dictionary<string, string> ProjectDisplayNames = new()
    {
        { "android-java", "Android Java" },
        { "android-kotlin", "Android Kotlin" },
        { "cpp-tui", "C++ TUI" },
        { "dotnet-maui", ".NET MAUI" },
        { "dotnet-tui", ".NET TUI" },
        { "dotnet-winforms", ".NET WinForms" },
        { "edge-server", "Edge Server" },
        { "flutter_app", "Flutter" },
        { "go-tui", "Go TUI" },
        { "java-server", "Java Server" },
        { "javascript-tui", "JavaScript TUI" },
        { "javascript-web", "JavaScript Web" },
        { "kotlin-multiplatform", "Kotlin Multiplatform" },
        { "react-native", "React Native" },
        { "react-native-expo", "React Native Expo" },
        { "rust-tui", "Rust TUI" },
        { "swift", "Swift" }
    };

    // MARK: - Download & Extract

    /// <summary>
    /// Downloads the quickstart zip from GitHub, extracts it to the given destination directory,
    /// and returns the path to the extracted quickstart-main folder.
    /// </summary>
    public async Task<string> DownloadAndExtractAsync(string destinationDirectory, IProgress<string>? progress = null)
    {
        progress?.Report("Starting download...");

        var tempZipPath = System.IO.Path.Combine(System.IO.Path.GetTempPath(), $"quickstart-{Guid.NewGuid()}.zip");
        var tempExtractDir = System.IO.Path.Combine(System.IO.Path.GetTempPath(), $"quickstart-extract-{Guid.NewGuid()}");

        try
        {
            // Download zip to temp file
            progress?.Report("Downloading quickstart zip...");
            using var httpClient = new HttpClient();
            var zipBytes = await httpClient.GetByteArrayAsync(ZipUrl);
            await File.WriteAllBytesAsync(tempZipPath, zipBytes);

            progress?.Report("Extracting archive...");

            // Extract to temp directory
            Directory.CreateDirectory(tempExtractDir);
            ZipFile.ExtractToDirectory(tempZipPath, tempExtractDir, overwriteFiles: true);

            // Copy the inner quickstart-main folder to the destination
            var extractedSourcePath = System.IO.Path.Combine(tempExtractDir, ExtractedFolderName);
            var destinationPath = System.IO.Path.Combine(destinationDirectory, ExtractedFolderName);

            if (Directory.Exists(destinationPath))
            {
                Directory.Delete(destinationPath, recursive: true);
            }

            CopyDirectory(extractedSourcePath, destinationPath);

            progress?.Report("Extraction complete.");
            return destinationPath;
        }
        finally
        {
            // Clean up temp files
            if (File.Exists(tempZipPath))
            {
                try { File.Delete(tempZipPath); } catch { /* best effort */ }
            }

            if (Directory.Exists(tempExtractDir))
            {
                try { Directory.Delete(tempExtractDir, recursive: true); } catch { /* best effort */ }
            }
        }
    }

    // MARK: - Configure .env Files

    /// <summary>
    /// Generates .env content and writes it to three locations:
    /// repo root, flutter_app/, and go-tui/.
    /// Skips non-existent subdirectories gracefully.
    /// </summary>
    public void ConfigureEnvFiles(
        string quickstartDir,
        string databaseId,
        string authToken,
        string authUrl,
        string websocketUrl)
    {
        var envContent =
            $"# Auto-configured by Edge Studio\n" +
            $"DITTO_APP_ID=\"{databaseId}\"\n" +
            $"DITTO_PLAYGROUND_TOKEN=\"{authToken}\"\n" +
            $"DITTO_AUTH_URL=\"{authUrl}\"\n" +
            $"DITTO_WEBSOCKET_URL=\"{websocketUrl}\"";

        var envLocations = new[]
        {
            System.IO.Path.Combine(quickstartDir, ".env"),
            System.IO.Path.Combine(quickstartDir, "flutter_app", ".env"),
            System.IO.Path.Combine(quickstartDir, "go-tui", ".env")
        };

        foreach (var envPath in envLocations)
        {
            var directory = System.IO.Path.GetDirectoryName(envPath);
            if (directory != null && Directory.Exists(directory))
            {
                File.WriteAllText(envPath, envContent);
            }
        }
    }

    // MARK: - Configure Edge Server YAML

    /// <summary>
    /// Generates quickstart_config.yaml in the edge-server/ subdirectory.
    /// </summary>
    public void ConfigureEdgeServerYaml(
        string quickstartDir,
        string databaseId,
        string authToken,
        string authUrl)
    {
        var yamlContent =
            $"resources:\n" +
            $"  my_ditto_db:\n" +
            $"    resource_type: DittoDatabase\n" +
            $"    db_id: \"{databaseId}\"\n" +
            $"    device_name: \"edge-studio-quickstart\"\n" +
            $"    subscriptions:\n" +
            $"      - \"SELECT * FROM tasks\"\n" +
            $"    auth:\n" +
            $"      server:\n" +
            $"        access_token: \"{authToken}\"\n" +
            $"        auth_url: \"{authUrl}\"\n" +
            $"        provider: \"__playgroundProvider\"\n" +
            $"  my_http_server:\n" +
            $"    resource_type: HttpServer\n" +
            $"    listen_addr: \"0.0.0.0:8080\"\n" +
            $"    databases:\n" +
            $"      my_db:\n" +
            $"        db_id: \"{databaseId}\"\n" +
            $"        base_path: my_server\n" +
            $"        http_api: true";

        var edgeServerDir = System.IO.Path.Combine(quickstartDir, "edge-server");
        if (!Directory.Exists(edgeServerDir))
        {
            return;
        }

        var yamlPath = System.IO.Path.Combine(edgeServerDir, "quickstart_config.yaml");
        File.WriteAllText(yamlPath, yamlContent);
    }

    // MARK: - Discover Projects

    /// <summary>
    /// Scans the quickstart directory for known project folders.
    /// Returns a list sorted by directory name.
    /// </summary>
    public List<QuickstartProject> DiscoverProjects(string quickstartDir, bool isConfigured)
    {
        var discovered = new List<QuickstartProject>();

        foreach (var (dirName, displayName) in ProjectDisplayNames)
        {
            var projectPath = System.IO.Path.Combine(quickstartDir, dirName);
            if (Directory.Exists(projectPath))
            {
                discovered.Add(new QuickstartProject(
                    Name: displayName,
                    DirectoryName: dirName,
                    Path: projectPath,
                    IsConfigured: isConfigured));
            }
        }

        return discovered.OrderBy(p => p.DirectoryName).ToList();
    }

    // MARK: - Existing Folder Check

    /// <summary>
    /// Returns the path to the quickstart-main folder if it exists inside the given directory,
    /// or null if it does not exist.
    /// </summary>
    public string? ExistingQuickstartFolder(string directory)
    {
        var folderPath = System.IO.Path.Combine(directory, ExtractedFolderName);
        return Directory.Exists(folderPath) ? folderPath : null;
    }

    // MARK: - Remove Folder

    /// <summary>
    /// Recursively removes the directory at the given path.
    /// </summary>
    public void RemoveExistingFolder(string path)
    {
        Directory.Delete(path, recursive: true);
    }

    // MARK: - Private Helpers

    /// <summary>
    /// Recursively copies a directory and all its contents to a new location.
    /// </summary>
    private static void CopyDirectory(string sourceDir, string destDir)
    {
        Directory.CreateDirectory(destDir);

        foreach (var file in Directory.GetFiles(sourceDir))
        {
            var destFile = System.IO.Path.Combine(destDir, System.IO.Path.GetFileName(file));
            File.Copy(file, destFile, overwrite: true);
        }

        foreach (var subDir in Directory.GetDirectories(sourceDir))
        {
            var destSubDir = System.IO.Path.Combine(destDir, System.IO.Path.GetFileName(subDir));
            CopyDirectory(subDir, destSubDir);
        }
    }
}
